module/zfs/arc.c

```diff
@@ -3472,7 +3473,12 @@ arc_prune_async(int64_t adjust)
 
 		refcount_add(&ap->p_refcnt, ap->p_pfunc);
 		ap->p_adjust = adjust;
-		taskq_dispatch(arc_prune_taskq, arc_prune_task, ap, TQ_SLEEP);
+		id = taskq_dispatch(arc_prune_taskq, arc_prune_task,
+			ap, TQ_SLEEP);
+		if (id == 0) {
+			refcount_remove(&ap->p_refcnt, ap->p_pfunc);
+			return;
+		}
```

Comment:
In this case you need to `continue` and not return.  It's entirely possible the dispatch will succeed for other entries in the list.  And you need to make sure you bump the counters and drop the mutex.  Something like this.

```
                if ((id = taskq_dispatch(arc_prune_taskq, arc_prune_task, 
                    ap, TQ_SLEEP)) == 0)
                        refcount_remove(&ap->p_refcnt, ap->p_pfunc);
```

Again, it probably best to break this in to its own PR for review/comment so it doesn't hold up these other trivial fixes.